src/pm/hydra/pm/pmiserv/pmip.c

```diff
@@ -181,10 +182,24 @@ int main(int argc, char **argv)
         }
         if (!count)
             break;
+
+        pid = waitpid(-1, &ret_status, WNOHANG);
+        if (pid > 0) {
```

Comment:
I'm concerned this check could be racy, meaning a failure could slip by and still end up in the loop below which causes the hang. Is there a way we can confirm that a failed process has not initialized PMI and therefore should trigger this cleanup phase?

Reply:
I'm not sure I understand what race you are referring to.  I'm only trying to catch cases where we don't exit this first loop at all.

Reply:
FYI, if the first process has initiated PMI and then aborts, we already check for that failure and clean up the processes (because we know that the socket closed abnormally).  Only in the case where the process aborts without initiating PMI that there's no such socket.  So, in that case we have to check for the exit status of the process and kill everything if the process exited with a signal.

Reply:
We resolved out of band. I didn't see that the check in question was happening inside a separate wait loop that ensures all children have responded first before waiting for termination below.